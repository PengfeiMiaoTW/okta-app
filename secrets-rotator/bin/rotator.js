"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.decryptEncryptedReportFile = exports.runSecretRotation = exports.getConfigData = exports.validateIfHashKeyEnvironmentVariableIsPresent = exports.substituteEnvironmentVariables = exports.rotateSecrets = void 0;
const types_1 = require("./types");
const db_password_rotator_1 = require("./gcp/db-password-rotator");
const generate_password_1 = require("generate-password");
const secret_manager_1 = require("./gcp/secret-manager");
const api_action_1 = require("./api/api-action");
const sendgrid_key_1 = require("./sendgrid/sendgrid-key");
const service_account_1 = require("./gcp/service-account");
const logger_1 = require("./logger");
const cast_to_1 = __importStar(require("./utils/cast-to"));
const fs_1 = require("fs");
const data_1 = require("./data");
const functions_1 = require("./utils/functions");
const encrypt_1 = require("./utils/encrypt");
const api_service_token_1 = require("./api-platform/api-service-token");
const redeploy_1 = require("./circleci/redeploy");
const event_subscription_token_1 = require("./event-platform/event-subscription-token");
const validate_schema_1 = require("./utils/validate-schema");
const secret_manager_2 = require("./aws/secret-manager");
const api_gateway_api_key_1 = require("./aws/api-gateway-api-key");
function generatePasswordAndPerformSecretRotation(source, handlerFunction) {
    return __awaiter(this, void 0, void 0, function* () {
        const password = yield generatePassword(source.passwordConfig);
        console.info("Password generated successfully");
        yield handlerFunction(source.action, password);
        return password;
    });
}
function generatePassword(passwordConfig) {
    return __awaiter(this, void 0, void 0, function* () {
        return (0, generate_password_1.generate)(passwordConfig);
    });
}
function performSourceAction(source) {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            console.info(`Rotation of source secret for "${source.name}" started`);
            const sourceActionHandlerMap = {
                [types_1.SourceActionType.GCP_POSTGRES_DATABASE]: () => generatePasswordAndPerformSecretRotation(source, db_password_rotator_1.rotatePostgresDbPassword),
                [types_1.SourceActionType.API]: () => (0, api_action_1.rotateSourceSecretByApiCall)((0, cast_to_1.default)(source.action)),
                [types_1.SourceActionType.API_PLATFORM_CLIENT_SECRET]: () => (0, api_action_1.rotateClientSecretByApiCall)((0, cast_to_1.default)(source.action)),
                [types_1.SourceActionType.API_PLATFORM_API_SERVICE_TOKEN]: () => generatePasswordAndPerformSecretRotation(source, api_service_token_1.rotateApiPlatformApiServiceToken),
                [types_1.SourceActionType.SENDGRID]: () => (0, sendgrid_key_1.rotateSendGridApiKey)((0, cast_to_1.default)(source.action)),
                [types_1.SourceActionType.GCP_SERVICE_ACCOUNT_KEY]: () => (0, service_account_1.rotateServiceAccount)((0, cast_to_1.default)(source.action)),
                [types_1.SourceActionType.GCP_SECRET_MANAGER]: () => generatePasswordAndPerformSecretRotation(source, secret_manager_1.rotateSecretsInSecretManager),
                [types_1.SourceActionType.AWS_SECRET_MANAGER]: () => generatePasswordAndPerformSecretRotation(source, secret_manager_2.rotateSecretsInAWSSecretManager),
                [types_1.SourceActionType.AWS_GATEWAY_API_KEY]: () => (0, api_gateway_api_key_1.rotateAWSGatewayApiKey)((0, cast_to_1.default)(source.action)),
                [types_1.SourceActionType.GENERATE_PASSWORD]: () => generatePassword(source.passwordConfig),
            };
            const password = yield sourceActionHandlerMap[source.action.type]();
            console.info(`Successfully rotated secret in source for "${source.name}"`);
            logger_1.reportFileLogger.appendText(`New secret value for "${source.name}": ${password}`);
            return password;
        }
        catch (error) {
            // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
            console.info(`Failed to rotate source secret for "${source.name}". Error Message: ${error.message}`);
            throw error;
        }
    });
}
function performConsumerAction(consumer, password) {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            console.info(`Rotation of secret for consumer: "${consumer.name}" started`);
            const consumerActionHandlerMap = {
                [types_1.ConsumerActionType.GCP_SECRET_MANAGER]: () => (0, secret_manager_1.rotateSecretsInSecretManager)((0, cast_to_1.default)(consumer.action), password),
                [types_1.ConsumerActionType.API]: () => (0, api_action_1.rotateConsumerSecretByApiCall)((0, cast_to_1.default)(consumer.action), password),
                [types_1.ConsumerActionType.API_PLATFORM_API_SERVICE_TOKEN]: () => (0, api_service_token_1.rotateApiPlatformApiServiceToken)((0, cast_to_1.default)(consumer.action), password),
                [types_1.ConsumerActionType.EVENT_SUBSCRIPTION_TOKEN]: () => (0, event_subscription_token_1.rotateEventSubscriptionToken)((0, cast_to_1.default)(consumer.action), password), [types_1.ConsumerActionType.AWS_SECRET_MANAGER]: () => (0, secret_manager_2.rotateSecretsInAWSSecretManager)((0, cast_to_1.default)(consumer.action), password), [types_1.ConsumerActionType.AWS_GATEWAY_API_KEY]: () => (0, api_gateway_api_key_1.rotateAWSGatewayApiKey)((0, cast_to_1.default)(consumer.action), password),
            };
            yield consumerActionHandlerMap[consumer.action.type]();
            console.info(`Successfully rotated secret of consumer: "${consumer.name}"`);
        }
        catch (error) {
            // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
            console.info(`Failed to rotate secret of consumer: "${consumer.name}". Error Message: ${error.message}`);
            throw error;
        }
    });
}
function rotateSecrets(config) {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            const { source, consumers } = config;
            const password = yield performSourceAction(source);
            for (const consumer of consumers) {
                yield performConsumerAction(consumer, password);
                if (consumer.redeploy) {
                    Array.isArray(consumer.redeploy)
                        ? consumer.redeploy.map((redeployConfig) => __awaiter(this, void 0, void 0, function* () {
                            yield (0, redeploy_1.redeploy)(redeployConfig);
                        }))
                        : yield (0, redeploy_1.redeploy)(consumer.redeploy);
                }
            }
        }
        catch (error) {
            console.info("Failed to rotate secret");
            // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
            console.error(error.message);
            // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
            throw new Error(error.message);
        }
    });
}
exports.rotateSecrets = rotateSecrets;
function substituteEnvironmentVariables(json) {
    const ENV_PATTERN = /\${([\w-_]*)}/gm;
    let matchersResult;
    let resultedJson = json;
    while ((matchersResult = ENV_PATTERN.exec(json)) !== null) {
        const [textToReplace, envVariableName] = matchersResult;
        // eslint-disable-next-line security/detect-object-injection
        if (!process.env[envVariableName]) {
            const errorMessage = `Environment variable: ${envVariableName} mentioned in config has no value.`;
            console.info(errorMessage);
            throw new Error(errorMessage);
        }
        // eslint-disable-next-line security/detect-object-injection
        resultedJson = resultedJson.replace(textToReplace, process.env[envVariableName]);
    }
    return resultedJson;
}
exports.substituteEnvironmentVariables = substituteEnvironmentVariables;
function validateIfHashKeyEnvironmentVariableIsPresent() {
    const hashKeyEnvVariableName = "HASH_KEY";
    // eslint-disable-next-line security/detect-object-injection
    if (!process.env[hashKeyEnvVariableName]) {
        const errorMessage = `Environment variable: ${hashKeyEnvVariableName} needed for secret rotation is not present.`;
        console.info(errorMessage);
        throw new Error(errorMessage);
    }
}
exports.validateIfHashKeyEnvironmentVariableIsPresent = validateIfHashKeyEnvironmentVariableIsPresent;
function readJsonFile(filePath) {
    return __awaiter(this, void 0, void 0, function* () {
        // eslint-disable-next-line security/detect-non-literal-fs-filename
        return fs_1.promises.readFile(filePath, "utf8");
    });
}
function getConfigData(filePath) {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            console.info("Started to read config file.");
            const jsonDataAsString = yield readJsonFile(filePath);
            const config = (0, cast_to_1.identity)(JSON.parse(substituteEnvironmentVariables(jsonDataAsString)));
            console.info("Successfully read config file.");
            return config;
        }
        catch (error) {
            console.info("Failed to read config json file.");
            throw error;
        }
    });
}
exports.getConfigData = getConfigData;
function runSecretRotation(configFilePath) {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            console.info("Started running secret rotation.");
            validateIfHashKeyEnvironmentVariableIsPresent();
            const config = yield getConfigData(configFilePath);
            yield (0, validate_schema_1.validateConfigFileSchema)(configFilePath);
            yield rotateSecrets(config);
            console.info("Secret rotation completed.");
        }
        catch (error) {
            logger_1.reportFileLogger.appendError("Failed to rotate secrets", error);
            throw error;
        }
        finally {
            console.info("Generating report file.");
            yield (0, functions_1.iif)(function generateEncryptedReportFile() {
                return __awaiter(this, void 0, void 0, function* () {
                    // eslint-disable-next-line security/detect-non-literal-fs-filename
                    const fileInBufferFormat = yield fs_1.promises.readFile(data_1.outputFileName);
                    // eslint-disable-next-line security/detect-non-literal-fs-filename
                    yield fs_1.promises.writeFile(data_1.encryptedReportFileName, (0, encrypt_1.encrypt)(fileInBufferFormat));
                });
            });
            yield (0, functions_1.iif)(function removeTemporaryReportFile() {
                return __awaiter(this, void 0, void 0, function* () {
                    // eslint-disable-next-line security/detect-non-literal-fs-filename
                    yield fs_1.promises.unlink(data_1.outputFileName);
                });
            });
            console.info("Report file generated successfully.");
        }
    });
}
exports.runSecretRotation = runSecretRotation;
function decryptEncryptedReportFile(encryptedReportFilePath) {
    return __awaiter(this, void 0, void 0, function* () {
        // eslint-disable-next-line security/detect-non-literal-fs-filename
        const fileContent = yield fs_1.promises.readFile(encryptedReportFilePath);
        // eslint-disable-next-line security/detect-non-literal-fs-filename
        yield fs_1.promises.writeFile(data_1.decryptedReportFileName, (0, encrypt_1.decrypt)(fileContent).toString());
    });
}
exports.decryptEncryptedReportFile = decryptEncryptedReportFile;
//# sourceMappingURL=rotator.js.map