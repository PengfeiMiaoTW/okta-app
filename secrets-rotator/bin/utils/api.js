"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCredentialsForBasicAuth = void 0;
const buffer_1 = require("buffer");
function getCredentialsForBasicAuth(config) {
    return buffer_1.Buffer.from(`${config.userName}:${config.password}`, "utf8").toString("base64");
}
exports.getCredentialsForBasicAuth = getCredentialsForBasicAuth;
//# sourceMappingURL=api.js.map