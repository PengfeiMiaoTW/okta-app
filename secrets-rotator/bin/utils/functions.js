"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.iif = exports.invokeImmediately = void 0;
function invokeImmediately(functionToInvoke) {
    return functionToInvoke();
}
exports.invokeImmediately = invokeImmediately;
exports.iif = invokeImmediately;
//# sourceMappingURL=functions.js.map