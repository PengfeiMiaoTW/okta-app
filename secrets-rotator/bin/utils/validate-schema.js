"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateConfigFileSchema = exports.validateJsonSchema = void 0;
const ajv_1 = __importDefault(require("ajv"));
const config_file_1 = require("../schema/config-file");
const rotator_1 = require("../rotator");
const ajv_errors_1 = __importDefault(require("ajv-errors"));
const ajv_formats_1 = __importDefault(require("ajv-formats"));
const uniqueItemProperties_1 = __importDefault(require("ajv-keywords/dist/definitions/uniqueItemProperties"));
const validateJsonSchema = (data) => {
    const schemaValidator = new ajv_1.default({
        allErrors: true,
    });
    (0, ajv_errors_1.default)(schemaValidator);
    schemaValidator.addKeyword((0, uniqueItemProperties_1.default)());
    schemaValidator.addKeyword("example");
    (0, ajv_formats_1.default)(schemaValidator);
    const validate = schemaValidator.compile(config_file_1.configFileSchema);
    if (!validate(data)) {
        throw new Error(JSON.stringify((validate.errors).map((error) => error.params.failingKeyword !== "then" ?
            (error.params.allowedValues ? `${error.message} ` + `${error.params.allowedValues} ` + "in " + `${error.instancePath} `
                // eslint-disable-next-line unicorn/no-nested-ternary
                : error.params.allowedValue ? `${error.message} ` + `${error.params.allowedValue} ` + "in " + `${error.instancePath} `
                    : `${error.message} ` + "in " + `${error.instancePath} `) : "").filter((errorMessage) => errorMessage)));
    }
    return true;
};
exports.validateJsonSchema = validateJsonSchema;
const validateConfigFileSchema = (configFilePath) => __awaiter(void 0, void 0, void 0, function* () {
    try {
        console.info("Validating config file schema");
        const configFileData = yield (0, rotator_1.getConfigData)(configFilePath);
        (0, exports.validateJsonSchema)(configFileData);
        console.info("Schema validation successful");
    }
    catch (error) {
        console.info("Schema validation failed");
        throw error;
    }
});
exports.validateConfigFileSchema = validateConfigFileSchema;
//# sourceMappingURL=validate-schema.js.map