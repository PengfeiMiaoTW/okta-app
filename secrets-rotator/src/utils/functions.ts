export function invokeImmediately<ReturnType>(functionToInvoke: () => ReturnType): ReturnType {
  return functionToInvoke();
}

export const iif = invokeImmediately;

