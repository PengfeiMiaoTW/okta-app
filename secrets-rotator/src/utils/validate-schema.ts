import Ajv from "ajv";
import { configFileSchema } from "../schema/config-file";
import { getConfigData } from "../rotator";
import addCustomErrorMessages from "ajv-errors";
import addFormats from "ajv-formats";
import getDef from "ajv-keywords/dist/definitions/uniqueItemProperties";


export const validateJsonSchema = (data: unknown) => {
  const schemaValidator = new Ajv({
    allErrors: true,
  });
  addCustomErrorMessages(schemaValidator);
  schemaValidator.addKeyword(getDef());
  schemaValidator.addKeyword("example");
  addFormats(schemaValidator);
  const validate = schemaValidator.compile(configFileSchema);
  if (!validate(data)) {
    throw new Error(JSON.stringify((validate.errors).map(
      (error) =>
        error.params.failingKeyword !== "then" ?
          (error.params.allowedValues ? `${error.message} ` + `${error.params.allowedValues} ` + "in " + `${error.instancePath} `
            // eslint-disable-next-line unicorn/no-nested-ternary
            : error.params.allowedValue ? `${error.message} ` + `${error.params.allowedValue} ` + "in " + `${error.instancePath} `
              : `${error.message} ` + "in " + `${error.instancePath} `) : "").filter((errorMessage) => errorMessage)));
  }
  return true;
};

export const validateConfigFileSchema = async (configFilePath: string) => {
  try {
    console.info("Validating config file schema");
    const configFileData = await getConfigData(configFilePath);
    validateJsonSchema(configFileData);
    console.info("Schema validation successful");
  } catch (error) {
    console.info("Schema validation failed");
    throw error;
  }
};
