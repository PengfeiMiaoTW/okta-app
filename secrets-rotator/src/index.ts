#!/usr/bin/env node
import yargs from "yargs";
import { decryptEncryptedReportFile, runSecretRotation } from "./rotator";
import { validateConfigFileSchema } from "./utils/validate-schema";

export function main() {
  // eslint-disable-next-line @typescript-eslint/no-unused-expressions
  yargs
    .command(
      "rotate",
      "Rotates secrets specified in the config file",
      (yarg) => yarg
        .option("config", {
          alias: "c",
          describe: "Configuration json file path",
          type: "string",
        })
        .demandOption(["config"], "Please provide the config json file path"),
      async function(argv) {
        await runSecretRotation(argv.config);
      },
    )
    .command(
      "validate-schema",
      "validate specified config file schema",
      (yarg) => yarg
        .option("config", {
          alias: "c",
          describe: "configuration json file path",
          type: "string",
        })
        .demandOption(["config"], "please provide the config json file path"),
      async function(argv) {
        await validateConfigFileSchema(argv.config);
      },
    )
    .command(
      "decrypt",
      "Decrypts encrypted report file",
      (yarg) => yarg
        .option("report", {
          alias: "r",
          describe: "Report file path",
          type: "string",
        })
        .demandOption(["report"], "Please provide the encrypted report file path"),
      async function(argv) {
        await decryptEncryptedReportFile(argv.report);
      },
    )
    .demandCommand()
    .help()
    .argv;
}

main();
