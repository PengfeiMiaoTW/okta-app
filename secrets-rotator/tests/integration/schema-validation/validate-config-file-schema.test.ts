import { iif } from "../../../src/utils/functions";
import "jest-extended";
import { validateJsonSchema } from "../../../src/utils/validate-schema";
import { getConfigData } from "../../../src/rotator";


const sendgridApiKeyName = "sendgrid-api-key-name";
const eventSubscriptionName = "application-event-subscription";
const eventSubscriptionAuthHeaderName = "test-subscription-auth-header";
const apiPlatformApiName = "application-service";
const oktaUrlToGenerateAccessTokenForApiGateway = "http://okta.com/access-token-from-client-id-client-secret";
const apiPlatformBaseUrl = "http://api-platform-base-url";
const apiPlatformConsumerClientId = "test-consumer-client-id";
const apiPlatformConsumerClientSecret = "test-consumer-client-secret";
const gcpProjectId = "12310283";
const gcpSecretName = "test-secret-name";
const apiPlatformAccessTokenGrantType = "test-grant";
const apiPlatformAccessTokenScope = "test-scope";
const gcloudKey = "ewogICJwcml2YXRlX2tleSI6ICIiLAogICJjbGllbnRfZW1haWwiOiAiIgp9Cg==";
const hashKey = "test-hash-key";
const sourceName = "Generate Password";
const dbUserName = "testuser";
const dbInstanceName = "test-db";
const gcpServiceAccountName = "svc-some-account@some-gcp-project.iam.gserviceaccount.com";
const circleCIUpdateEnvironmentVariableUrl = "https://circleci.com/api/v2/context/context-id/environment-variable/GCLOUD_KEY";
const circleCiApiToken = "test-circleci-api-token";
const oldSendGridApiKey = "old-sendgrid-api-key";
const workflowId1 = "workflow-000001";
const jobId1 = "job-000001";
const circleCIAccessToken = "test-token";
const secretName = "test-secret-name";
const usagePlanId = "test-usage-plan-id";
const region = "sample-region";


beforeEach(() => {
  iif(function setupEnv() {
    process.env = {
      ...process.env,
      SOURCE_NAME: sourceName,
      HASH_KEY: hashKey,
      API_PLATFORM_API_NAME: apiPlatformApiName,
      API_PLATFORM_TOKEN_URL: oktaUrlToGenerateAccessTokenForApiGateway,
      API_PLATFORM_URL: apiPlatformBaseUrl,
      API_PLATFORM_CLIENT_ID: apiPlatformConsumerClientId,
      API_PLATFORM_CLIENT_SECRET: apiPlatformConsumerClientSecret,
      API_PLATFORM_ACCESS_TOKEN_GRANT_TYPE: apiPlatformAccessTokenGrantType,
      API_PLATFORM_ACCESS_TOKEN_SCOPE: apiPlatformAccessTokenScope,
      GCP_PROJECT_ID: gcpProjectId,
      GCP_SECRET_NAME: gcpSecretName,
      GCLOUD_KEY: gcloudKey,
      WORKFLOW_ID: workflowId1,
      JOB_ID: jobId1,
      WORKFLOW_ID_SERVICE_TWO: workflowId1,
      JOB_ID_SERVICE_TWO: jobId1,
      CIRCLE_CI_ACCESS_TOKEN: circleCIAccessToken,
      DB_USERNAME: dbUserName,
      DB_INSTANCE_NAME_FOR_SECRET_ROTATION: dbInstanceName,
      EVENT_SUBSCRIPTION_NAME: eventSubscriptionName,
      EVENT_SUBSCRIPTION_AUTH_HEADER_NAME: eventSubscriptionAuthHeaderName,
      SENDGRID_API_KEY_NAME: sendgridApiKeyName,
      GCP_SERVICE_ACCOUNT_NAME: gcpServiceAccountName,
      CIRCLECI_CONTEXT_ENV_VARIABLE_UPDATE_URL: circleCIUpdateEnvironmentVariableUrl,
      CIRCLECI_API_TOKEN: circleCiApiToken,
      SENDGRID_API_KEY: oldSendGridApiKey,
      AWS_SECRET_NAME: secretName,
      USAGE_PLAN_ID: usagePlanId,
      REGION: region,
    };
  });
  jest.useFakeTimers();
});


describe("configuration files schema validation", () => {
  describe("schema validation successful scenarios", () => {

    it.each([
      "./tests/integration/gcp/mock-configs/api-platform-api-token-rotation-config-for-secrets-in-json-format.json",
      "./tests/integration/gcp/mock-configs/api-platform-api-token-rotation-config-for-secrets-with-redeploy-array-in-json-format.json",
      "./tests/integration/gcp/mock-configs/api-platform-api-token-rotation-config-for-secrets-in-plain-text-format.json",
      "./tests/integration/gcp/mock-configs/api-platform-api-token-rotation-config-for-secrets-with-keyFieldNames-in-json-format.json",
      "./tests/integration/gcp/mock-configs/app-session-secret-rotation-config.json",
      "./tests/integration/gcp/mock-configs/cda-client-secret-rotation-config-when-secret-is-stored-in-json-format.json",
      "./tests/integration/gcp/mock-configs/cda-client-secret-rotation-when-secret-is-stored-in-plain-text-in-format.json",
      "./tests/integration/gcp/mock-configs/cda-client-secret-rotation-config-with-keyFieldNames-when-secret-is-stored-in-json-format.json",
      "./tests/integration/gcp/mock-configs/db-rotation-config.json",
      "./tests/integration/gcp/mock-configs/event-subscription-token-rotation-config-for-secrets-in-json-format.json",
      "./tests/integration/gcp/mock-configs/event-subscription-token-rotation-config-for-secrets-in-plain-text-format.json",
      "./tests/integration/gcp/mock-configs/event-subscription-token-rotation-config-for-secrets-with-keyFieldNames-in-json-format.json",
      "./tests/integration/gcp/mock-configs/sendgrid-key-rotation-config.json",
      "./tests/integration/gcp/mock-configs/service-account-rotation-config.json",
      "./tests/integration/aws/mock-configs/static-token-rotation.json",
      "./tests/integration/aws/mock-configs/cda-client-secret-rotation-config-when-secret-is-stored-as-array-in-json-format.json",
      "./tests/integration/aws/mock-configs/cda-client-secret-rotation-config-with-keyFieldNames-when-secret-is-stored-in-json-format.json",
      "./tests/integration/aws/mock-configs/valid-aws-api-gateway-api-key-rotation-config.json",
      "./tests/integration/aws/mock-configs/aws-api-gateway-api-key-rotation-config-as-source-action.json",
    ])("should not throw error given valid config file", async (configFilePath) => {
      const configFileData = await getConfigData(configFilePath);
      expect(
        () => validateJsonSchema(configFileData),
      ).not.toThrow();
    });
  });

  describe("schema validation failure scenarios", () => {
    it("should throw error for invalid gcp config file", async () => {
      const configFileData = await getConfigData("./tests/integration/gcp/mock-configs/app-session-secret-rotation-invalid-config.json");
      expect(
        () => validateJsonSchema(configFileData),
      ).toThrowErrorMatchingSnapshot();
    });

    it.each([
      "./tests/integration/aws/mock-configs/invalid-aws-secret-manager-secret-rotation-config.json",
      "./tests/integration/aws/mock-configs/invalid-aws-api-gateway-api-key-rotation-config.json",
    ])("should throw error for invalid aws config files", async (configFilePath) => {
      const configFileData = await getConfigData(configFilePath);
      expect(
        () => validateJsonSchema(configFileData),
      ).toThrowErrorMatchingSnapshot();
    });
  });
});
