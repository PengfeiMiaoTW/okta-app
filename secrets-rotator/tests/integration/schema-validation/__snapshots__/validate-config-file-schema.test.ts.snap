// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`configuration files schema validation schema validation failure scenarios should throw error for invalid aws config files 1`] = `"[\\"must have required property 'secretId' in /consumers/0/action \\",\\"must have required property 'region' in /consumers/0/action \\",\\"must have required property 'secretAccessKey' in /consumers/0/action/authentication/credentials \\"]"`;

exports[`configuration files schema validation schema validation failure scenarios should throw error for invalid aws config files 2`] = `"[\\"must have required property 'secretAccessKey' in /consumers/0/action/authentication/credentials \\"]"`;

exports[`configuration files schema validation schema validation failure scenarios should throw error for invalid gcp config file 1`] = `"[\\"must have required property 'name' in /source \\"]"`;
